Android remove backward timezones

mdh1418 opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/63693

It was discovered that Android produces duplicate TimeZone DisplayNames among all timezone IDs in `GetSystemTimeZones`. These duplicate DisplayNames occur across TimeZone IDs that are aliases, where all except one are backward timezone IDs. 
> If a name is changed, put its old spelling in the 'backward' file

From the Android TimeZone data file `tzdata`, it isn't obvious which TimeZone IDs are backward (I find it strange that they're included in the first place), however we discovered that on some versions of Android, there is an adjacent file `tzlookup.xml` that can aid us in determining which TimeZone IDs are "current" (not backward).

This PR aims to utilize `tzlookup.xml` when it exists and post-filter's the Populated TimeZone IDs in the `AndroidTzData` instance by removing IDs and their associated information (byteoffset and length) from the `AndroidTzData` instance if it is not found in `tzlookup.xml`. This is using the assumption that all non-backward TimeZone IDs make it to the `tzlookup.xml` file.

This PR also adds a new TimeZoneInfo Test to check whether or not there are duplicate DisplayNames in `GetSystemTimeZones`

---
**Android IDs and DisplayNames after Change**
<details>
    <summary>(click to expand)</summary>

| Id                             | Display Name                                         |
| ------------------------------ | ---------------------------------------------------- |
| Africa/Abidjan | (UTC+00:00) Côte d’Ivoire Time |
| Africa/Accra | (UTC+00:00) Ghana Time |
| Africa/Addis_Ababa | (UTC+03:00) East Africa Time (Addis Ababa) |
| Africa/Algiers | (UTC+01:00) Central European Time (Algiers) |
| Africa/Asmara | (UTC+03:00) East Africa Time (Asmara) |
| Africa/Bamako | (UTC+00:00) Mali Time |
| Africa/Bangui | (UTC+01:00) West Africa Standard Time (Bangui) |
| Africa/Banjul | (UTC+00:00) Gambia Time |
| Africa/Bissau | (UTC+00:00) Guinea-Bissau Time |
| Africa/Blantyre | (UTC+02:00) Central Africa Time (Blantyre) |
| Africa/Brazzaville | (UTC+01:00) West Africa Standard Time (Brazzaville) |
| Africa/Bujumbura | (UTC+02:00) Central Africa Time (Bujumbura) |
| Africa/Cairo | (UTC+02:00) Eastern European Time (Cairo) |
| Africa/Casablanca | (UTC+00:00) Morocco Time |
| Africa/Ceuta | (UTC+01:00) Central European Time (Ceuta) |
| Africa/Conakry | (UTC+00:00) Guinea Time |
| Africa/Dakar | (UTC+00:00) Senegal Time |
| Africa/Dar_es_Salaam | (UTC+03:00) East Africa Time (Dar es Salaam) |
| Africa/Djibouti | (UTC+03:00) East Africa Time (Djibouti) |
| Africa/Douala | (UTC+01:00) West Africa Standard Time (Douala) |
| Africa/El_Aaiun | (UTC+00:00) Western Sahara Time |
| Africa/Freetown | (UTC+00:00) Sierra Leone Time |
| Africa/Gaborone | (UTC+02:00) Central Africa Time (Gaborone) |
| Africa/Harare | (UTC+02:00) Central Africa Time (Harare) |
| Africa/Johannesburg | (UTC+02:00) South Africa Standard Time (Johannesburg) |
| Africa/Juba | (UTC+03:00) East Africa Time (Juba) |
| Africa/Kampala | (UTC+03:00) East Africa Time (Kampala) |
| Africa/Khartoum | (UTC+02:00) Central Africa Time (Khartoum) |
| Africa/Kigali | (UTC+02:00) Central Africa Time (Kigali) |
| Africa/Kinshasa | (UTC+01:00) West Africa Standard Time (Kinshasa) |
| Africa/Lagos | (UTC+01:00) West Africa Standard Time (Lagos) |
| Africa/Libreville | (UTC+01:00) West Africa Standard Time (Libreville) |
| Africa/Lome | (UTC+00:00) Togo Time |
| Africa/Luanda | (UTC+01:00) West Africa Standard Time (Luanda) |
| Africa/Lubumbashi | (UTC+02:00) Central Africa Time (Lubumbashi) |
| Africa/Lusaka | (UTC+02:00) Central Africa Time (Lusaka) |
| Africa/Malabo | (UTC+01:00) West Africa Standard Time (Malabo) |
| Africa/Maputo | (UTC+02:00) Central Africa Time (Maputo) |
| Africa/Maseru | (UTC+02:00) South Africa Standard Time (Maseru) |
| Africa/Mbabane | (UTC+02:00) South Africa Standard Time (Mbabane) |
| Africa/Mogadishu | (UTC+03:00) East Africa Time (Mogadishu) |
| Africa/Monrovia | (UTC+00:00) Liberia Time |
| Africa/Nairobi | (UTC+03:00) East Africa Time (Nairobi) |
| Africa/Ndjamena | (UTC+01:00) West Africa Standard Time (Ndjamena) |
| Africa/Niamey | (UTC+01:00) West Africa Standard Time (Niamey) |
| Africa/Nouakchott | (UTC+00:00) Mauritania Time |
| Africa/Ouagadougou | (UTC+00:00) Burkina Faso Time |
| Africa/Porto-Novo | (UTC+01:00) West Africa Standard Time (Porto-Novo) |
| Africa/Sao_Tome | (UTC+00:00) São Tomé & Príncipe Time |
| Africa/Tripoli | (UTC+02:00) Eastern European Time (Tripoli) |
| Africa/Tunis | (UTC+01:00) Central European Time (Tunis) |
| Africa/Windhoek | (UTC+02:00) Central Africa Time (Windhoek) |
| America/Adak | (UTC-10:00) Hawaii-Aleutian Time (Adak) |
| America/Anchorage | (UTC-09:00) Alaska Time (Anchorage) |
| America/Anguilla | (UTC-04:00) Atlantic Time (Anguilla) |
| America/Antigua | (UTC-04:00) Atlantic Time (Antigua) |
| America/Araguaina | (UTC-03:00) Brasilia Standard Time (Araguaina) |
| America/Argentina/Buenos_Aires | (UTC-03:00) Argentina Standard Time (Buenos Aires) |
| America/Argentina/Catamarca | (UTC-03:00) Argentina Standard Time (Catamarca) |
| America/Argentina/Cordoba | (UTC-03:00) Argentina Standard Time (Cordoba) |
| America/Argentina/Jujuy | (UTC-03:00) Argentina Standard Time (Jujuy) |
| America/Argentina/La_Rioja | (UTC-03:00) Argentina Standard Time (La Rioja) |
| America/Argentina/Mendoza | (UTC-03:00) Argentina Standard Time (Mendoza) |
| America/Argentina/Rio_Gallegos | (UTC-03:00) Argentina Standard Time (Rio Gallegos) |
| America/Argentina/Salta | (UTC-03:00) Argentina Standard Time (Salta) |
| America/Argentina/San_Juan | (UTC-03:00) Argentina Standard Time (San Juan) |
| America/Argentina/San_Luis | (UTC-03:00) Western Argentina Standard Time (San Luis) |
| America/Argentina/Tucuman | (UTC-03:00) Argentina Standard Time (Tucuman) |
| America/Argentina/Ushuaia | (UTC-03:00) Argentina Standard Time (Ushuaia) |
| America/Aruba | (UTC-04:00) Atlantic Time (Aruba) |
| America/Asuncion | (UTC-04:00) Paraguay Time |
| America/Atikokan | (UTC-05:00) Eastern Time (Atikokan) |
| America/Bahia | (UTC-03:00) Brasilia Standard Time (Bahia) |
| America/Bahia_Banderas | (UTC-06:00) Central Time (Bahia Banderas) |
| America/Barbados | (UTC-04:00) Atlantic Time (Barbados) |
| America/Belem | (UTC-03:00) Brasilia Standard Time (Belem) |
| America/Belize | (UTC-06:00) Central Time (Belize) |
| America/Blanc-Sablon | (UTC-04:00) Atlantic Time (Blanc-Sablon) |
| America/Boa_Vista | (UTC-04:00) Amazon Standard Time (Boa Vista) |
| America/Bogota | (UTC-05:00) Colombia Standard Time (Bogota) |
| America/Boise | (UTC-07:00) Mountain Time (Boise) |
| America/Cambridge_Bay | (UTC-07:00) Mountain Time (Cambridge Bay) |
| America/Campo_Grande | (UTC-04:00) Amazon Standard Time (Campo Grande) |
| America/Cancun | (UTC-05:00) Eastern Time (Cancun) |
| America/Caracas | (UTC-04:00) Venezuela Time |
| America/Cayenne | (UTC-03:00) French Guiana Time |
| America/Cayman | (UTC-05:00) Eastern Time (Cayman) |
| America/Chicago | (UTC-06:00) Central Time (Chicago) |
| America/Chihuahua | (UTC-07:00) Mexican Pacific Time (Chihuahua) |
| America/Costa_Rica | (UTC-06:00) Central Time (Costa Rica) |
| America/Creston | (UTC-07:00) Mountain Time (Creston) |
| America/Cuiaba | (UTC-04:00) Amazon Standard Time (Cuiaba) |
| America/Curacao | (UTC-04:00) Atlantic Time (Curaçao) |
| America/Danmarkshavn | (UTC+00:00) Danmarkshavn Time |
| America/Dawson | (UTC-08:00) Pacific Time (Dawson) |
| America/Dawson_Creek | (UTC-07:00) Mountain Time (Dawson Creek) |
| America/Denver | (UTC-07:00) Mountain Time (Denver) |
| America/Detroit | (UTC-05:00) Eastern Time (Detroit) |
| America/Dominica | (UTC-04:00) Atlantic Time (Dominica) |
| America/Edmonton | (UTC-07:00) Mountain Time (Edmonton) |
| America/Eirunepe | (UTC-05:00) Acre Standard Time (Eirunepe) |
| America/El_Salvador | (UTC-06:00) Central Time (El Salvador) |
| America/Fort_Nelson | (UTC-07:00) Mountain Time (Fort Nelson) |
| America/Fortaleza | (UTC-03:00) Brasilia Standard Time (Fortaleza) |
| America/Glace_Bay | (UTC-04:00) Atlantic Time (Glace Bay) |
| America/Godthab | (UTC-03:00) West Greenland Time (Nuuk) |
| America/Goose_Bay | (UTC-04:00) Atlantic Time (Goose Bay) |
| America/Grand_Turk | (UTC-05:00) Eastern Time (Grand Turk) |
| America/Grenada | (UTC-04:00) Atlantic Time (Grenada) |
| America/Guadeloupe | (UTC-04:00) Atlantic Time (Guadeloupe) |
| America/Guatemala | (UTC-06:00) Central Time (Guatemala) |
| America/Guayaquil | (UTC-05:00) Ecuador Time |
| America/Guyana | (UTC-04:00) Guyana Time |
| America/Halifax | (UTC-04:00) Atlantic Time (Halifax) |
| America/Havana | (UTC-05:00) Cuba Time |
| America/Hermosillo | (UTC-07:00) Mexican Pacific Time (Hermosillo) |
| America/Indiana/Indianapolis | (UTC-05:00) Eastern Time (Indianapolis) |
| America/Indiana/Knox | (UTC-06:00) Central Time (Knox, Indiana) |
| America/Indiana/Marengo | (UTC-05:00) Eastern Time (Marengo, Indiana) |
| America/Indiana/Petersburg | (UTC-05:00) Eastern Time (Petersburg, Indiana) |
| America/Indiana/Tell_City | (UTC-06:00) Central Time (Tell City, Indiana) |
| America/Indiana/Vevay | (UTC-05:00) Eastern Time (Vevay, Indiana) |
| America/Indiana/Vincennes | (UTC-05:00) Eastern Time (Vincennes, Indiana) |
| America/Indiana/Winamac | (UTC-05:00) Eastern Time (Winamac, Indiana) |
| America/Inuvik | (UTC-07:00) Mountain Time (Inuvik) |
| America/Iqaluit | (UTC-05:00) Eastern Time (Iqaluit) |
| America/Jamaica | (UTC-05:00) Eastern Time (Jamaica) |
| America/Juneau | (UTC-09:00) Alaska Time (Juneau) |
| America/Kentucky/Louisville | (UTC-05:00) Eastern Time (Louisville) |
| America/Kentucky/Monticello | (UTC-05:00) Eastern Time (Monticello, Kentucky) |
| America/Kralendijk | (UTC-04:00) Atlantic Time (Kralendijk) |
| America/La_Paz | (UTC-04:00) Bolivia Time |
| America/Lima | (UTC-05:00) Peru Standard Time (Lima) |
| America/Los_Angeles | (UTC-08:00) Pacific Time (Los Angeles) |
| America/Lower_Princes | (UTC-04:00) Atlantic Time (Lower Prince’s Quarter) |
| America/Maceio | (UTC-03:00) Brasilia Standard Time (Maceio) |
| America/Managua | (UTC-06:00) Central Time (Managua) |
| America/Manaus | (UTC-04:00) Amazon Standard Time (Manaus) |
| America/Marigot | (UTC-04:00) Atlantic Time (Marigot) |
| America/Martinique | (UTC-04:00) Atlantic Time (Martinique) |
| America/Matamoros | (UTC-06:00) Central Time (Matamoros) |
| America/Mazatlan | (UTC-07:00) Mexican Pacific Time (Mazatlan) |
| America/Menominee | (UTC-06:00) Central Time (Menominee) |
| America/Merida | (UTC-06:00) Central Time (Merida) |
| America/Metlakatla | (UTC-09:00) Alaska Time (Metlakatla) |
| America/Mexico_City | (UTC-06:00) Central Time (Mexico City) |
| America/Miquelon | (UTC-03:00) St. Pierre & Miquelon Time |
| America/Moncton | (UTC-04:00) Atlantic Time (Moncton) |
| America/Monterrey | (UTC-06:00) Central Time (Monterrey) |
| America/Montevideo | (UTC-03:00) Uruguay Standard Time (Montevideo) |
| America/Montserrat | (UTC-04:00) Atlantic Time (Montserrat) |
| America/Nassau | (UTC-05:00) Eastern Time (Nassau) |
| America/New_York | (UTC-05:00) Eastern Time (New York) |
| America/Nipigon | (UTC-05:00) Eastern Time (Nipigon) |
| America/Nome | (UTC-09:00) Alaska Time (Nome) |
| America/Noronha | (UTC-02:00) Fernando de Noronha Standard Time |
| America/North_Dakota/Beulah | (UTC-06:00) Central Time (Beulah, North Dakota) |
| America/North_Dakota/Center | (UTC-06:00) Central Time (Center, North Dakota) |
| America/North_Dakota/New_Salem | (UTC-06:00) Central Time (New Salem, North Dakota) |
| - | - |
| America/Ojinaga | (UTC-07:00) Mountain Time (Ojinaga) |
| America/Panama | (UTC-05:00) Eastern Time (Panama) |
| America/Pangnirtung | (UTC-05:00) Eastern Time (Pangnirtung) |
| America/Paramaribo | (UTC-03:00) Suriname Time |
| America/Phoenix | (UTC-07:00) Mountain Time (Phoenix) |
| America/Port-au-Prince | (UTC-05:00) Eastern Time (Port-au-Prince) |
| America/Port_of_Spain | (UTC-04:00) Atlantic Time (Port of Spain) |
| America/Porto_Velho | (UTC-04:00) Amazon Standard Time (Porto Velho) |
| America/Puerto_Rico | (UTC-04:00) Atlantic Time (Puerto Rico) |
| America/Punta_Arenas | (UTC-03:00) Punta Arenas Time |
| America/Rainy_River | (UTC-06:00) Central Time (Rainy River) |
| America/Rankin_Inlet | (UTC-06:00) Central Time (Rankin Inlet) |
| America/Recife | (UTC-03:00) Brasilia Standard Time (Recife) |
| America/Regina | (UTC-06:00) Central Time (Regina) |
| America/Resolute | (UTC-06:00) Central Time (Resolute) |
| America/Rio_Branco | (UTC-05:00) Acre Standard Time (Rio Branco) |
| America/Santarem | (UTC-03:00) Brasilia Standard Time (Santarem) |
| America/Santiago | (UTC-04:00) Chile Time |
| America/Santo_Domingo | (UTC-04:00) Atlantic Time (Santo Domingo) |
| America/Sao_Paulo | (UTC-03:00) Brasilia Standard Time (Sao Paulo) |
| America/Scoresbysund | (UTC-01:00) East Greenland Time (Ittoqqortoormiit) |
| America/Sitka | (UTC-09:00) Alaska Time (Sitka) |
| America/St_Barthelemy | (UTC-04:00) Atlantic Time (St. Barthélemy) |
| America/St_Johns | (UTC-03:30) Newfoundland Time (St. John’s) |
| America/St_Kitts | (UTC-04:00) Atlantic Time (St. Kitts) |
| America/St_Lucia | (UTC-04:00) Atlantic Time (St. Lucia) |
| America/St_Thomas | (UTC-04:00) Atlantic Time (St. Thomas) |
| America/St_Vincent | (UTC-04:00) Atlantic Time (St. Vincent) |
| America/Swift_Current | (UTC-06:00) Central Time (Swift Current) |
| America/Tegucigalpa | (UTC-06:00) Central Time (Tegucigalpa) |
| America/Thule | (UTC-04:00) Atlantic Time (Thule) |
| America/Thunder_Bay | (UTC-05:00) Eastern Time (Thunder Bay) |
| America/Tijuana | (UTC-08:00) Pacific Time (Tijuana) |
| America/Toronto | (UTC-05:00) Eastern Time (Toronto) |
| America/Tortola | (UTC-04:00) Atlantic Time (Tortola) |
| America/Vancouver | (UTC-08:00) Pacific Time (Vancouver) |
| America/Whitehorse | (UTC-08:00) Pacific Time (Whitehorse) |
| America/Winnipeg | (UTC-06:00) Central Time (Winnipeg) |
| America/Yakutat | (UTC-09:00) Alaska Time (Yakutat) |
| America/Yellowknife | (UTC-07:00) Mountain Time (Yellowknife) |
| Antarctica/Casey | (UTC+08:00) Australian Western Standard Time (Casey) |
| Antarctica/Davis | (UTC+07:00) Davis Time |
| Antarctica/DumontDUrville | (UTC+10:00) Dumont-d’Urville Time (Dumont d’Urville) |
| Antarctica/Macquarie | (UTC+11:00) Macquarie Island Time |
| Antarctica/Mawson | (UTC+05:00) Mawson Time |
| Antarctica/McMurdo | (UTC+12:00) New Zealand Time (McMurdo) |
| Antarctica/Palmer | (UTC-03:00) Palmer Time |
| Antarctica/Rothera | (UTC-03:00) Rothera Time |
| Antarctica/Syowa | (UTC+03:00) Syowa Time |
| Antarctica/Troll | (UTC+00:00) Troll Time |
| Antarctica/Vostok | (UTC+06:00) Vostok Time |
| Arctic/Longyearbyen | (UTC+01:00) Central European Time (Longyearbyen) |
| Asia/Aden | (UTC+03:00) Arabian Standard Time (Aden) |
| Asia/Almaty | (UTC+06:00) East Kazakhstan Time (Almaty) |
| Asia/Amman | (UTC+02:00) Eastern European Time (Amman) |
| Asia/Anadyr | (UTC+12:00) Anadyr Standard Time |
| Asia/Aqtau | (UTC+05:00) West Kazakhstan Time (Aqtau) |
| Asia/Aqtobe | (UTC+05:00) West Kazakhstan Time (Aqtobe) |
| Asia/Ashgabat | (UTC+05:00) Turkmenistan Standard Time (Ashgabat) |
| Asia/Atyrau | (UTC+05:00) West Kazakhstan Time (Atyrau) |
| Asia/Baghdad | (UTC+03:00) Arabian Standard Time (Baghdad) |
| Asia/Bahrain | (UTC+03:00) Arabian Standard Time (Bahrain) |
| Asia/Baku | (UTC+04:00) Azerbaijan Standard Time (Baku) |
| Asia/Bangkok | (UTC+07:00) Indochina Time (Bangkok) |
| Asia/Barnaul | (UTC+07:00) Barnaul Time |
| Asia/Beirut | (UTC+02:00) Eastern European Time (Beirut) |
| Asia/Bishkek | (UTC+06:00) Kyrgyzstan Time |
| Asia/Brunei | (UTC+08:00) Brunei Darussalam Time |
| Asia/Chita | (UTC+09:00) Yakutsk Standard Time (Chita) |
| Asia/Choibalsan | (UTC+08:00) Choibalsan Standard Time |
| Asia/Colombo | (UTC+05:30) India Standard Time (Colombo) |
| Asia/Damascus | (UTC+02:00) Eastern European Time (Damascus) |
| Asia/Dhaka | (UTC+06:00) Bangladesh Standard Time (Dhaka) |
| Asia/Dili | (UTC+09:00) East Timor Time (Dili) |
| Asia/Dubai | (UTC+04:00) Gulf Standard Time (Dubai) |
| Asia/Dushanbe | (UTC+05:00) Tajikistan Time |
| Asia/Famagusta | (UTC+02:00) Famagusta Time |
| Asia/Gaza | (UTC+02:00) Eastern European Time (Gaza) |
| Asia/Hebron | (UTC+02:00) Eastern European Time (Hebron) |
| Asia/Ho_Chi_Minh | (UTC+07:00) Indochina Time (Ho Chi Minh City) |
| Asia/Hong_Kong | (UTC+08:00) Hong Kong Standard Time |
| Asia/Hovd | (UTC+07:00) Hovd Standard Time |
| Asia/Irkutsk | (UTC+08:00) Irkutsk Standard Time |
| Asia/Jakarta | (UTC+07:00) Western Indonesia Time (Jakarta) |
| Asia/Jayapura | (UTC+09:00) Eastern Indonesia Time (Jayapura) |
| Asia/Jerusalem | (UTC+02:00) Israel Time |
| Asia/Kabul | (UTC+04:30) Afghanistan Time |
| Asia/Kamchatka | (UTC+12:00) Petropavlovsk-Kamchatski Standard Time (Kamchatka) |
| Asia/Karachi | (UTC+05:00) Pakistan Standard Time (Karachi) |
| Asia/Kathmandu | (UTC+05:45) Nepal Time |
| Asia/Khandyga | (UTC+09:00) Yakutsk Standard Time (Khandyga) |
| Asia/Kolkata | (UTC+05:30) India Standard Time (Kolkata) |
| Asia/Krasnoyarsk | (UTC+07:00) Krasnoyarsk Standard Time |
| Asia/Kuala_Lumpur | (UTC+08:00) Malaysia Time |
| Asia/Kuching | (UTC+08:00) Malaysia Time (Kuching) |
| Asia/Kuwait | (UTC+03:00) Arabian Standard Time (Kuwait) |
| Asia/Macau | (UTC+08:00) China Standard Time (Macau) |
| Asia/Magadan | (UTC+11:00) Magadan Standard Time |
| Asia/Makassar | (UTC+08:00) Central Indonesia Time (Makassar) |
| Asia/Manila | (UTC+08:00) Philippine Standard Time (Manila) |
| Asia/Muscat | (UTC+04:00) Gulf Standard Time (Muscat) |
| Asia/Nicosia | (UTC+02:00) Eastern European Time (Nicosia) |
| Asia/Novokuznetsk | (UTC+07:00) Krasnoyarsk Standard Time (Novokuznetsk) |
| Asia/Novosibirsk | (UTC+07:00) Novosibirsk Standard Time |
| Asia/Omsk | (UTC+06:00) Omsk Standard Time |
| Asia/Oral | (UTC+05:00) West Kazakhstan Time (Oral) |
| Asia/Phnom_Penh | (UTC+07:00) Indochina Time (Phnom Penh) |
| Asia/Pontianak | (UTC+07:00) Western Indonesia Time (Pontianak) |
| Asia/Pyongyang | (UTC+09:00) Korean Standard Time (Pyongyang) |
| Asia/Qatar | (UTC+03:00) Arabian Standard Time (Qatar) |
| Asia/Qostanay | (UTC+06:00) East Kazakhstan Time (Kostanay) |
| Asia/Qyzylorda | (UTC+05:00) West Kazakhstan Time (Qyzylorda) |
| Asia/Riyadh | (UTC+03:00) Arabian Standard Time (Riyadh) |
| Asia/Sakhalin | (UTC+11:00) Sakhalin Standard Time |
| Asia/Samarkand | (UTC+05:00) Uzbekistan Standard Time (Samarkand) |
| Asia/Seoul | (UTC+09:00) Korean Standard Time (Seoul) |
| Asia/Shanghai | (UTC+08:00) China Standard Time (Shanghai) |
| Asia/Singapore | (UTC+08:00) Singapore Standard Time |
| Asia/Srednekolymsk | (UTC+11:00) Srednekolymsk Time |
| Asia/Taipei | (UTC+08:00) Taipei Standard Time |
| Asia/Tashkent | (UTC+05:00) Uzbekistan Standard Time (Tashkent) |
| Asia/Tbilisi | (UTC+04:00) Georgia Standard Time (Tbilisi) |
| Asia/Tehran | (UTC+03:30) Iran Time |
| Asia/Thimphu | (UTC+06:00) Bhutan Time |
| Asia/Tokyo | (UTC+09:00) Japan Standard Time (Tokyo) |
| Asia/Tomsk | (UTC+07:00) Tomsk Time |
| Asia/Ulaanbaatar | (UTC+08:00) Ulaanbaatar Standard Time |
| Asia/Urumqi | (UTC+06:00) Urumqi Time |
| Asia/Ust-Nera | (UTC+10:00) Vladivostok Standard Time (Ust-Nera) |
| Asia/Vientiane | (UTC+07:00) Indochina Time (Vientiane) |
| Asia/Vladivostok | (UTC+10:00) Vladivostok Standard Time |
| Asia/Yakutsk | (UTC+09:00) Yakutsk Standard Time |
| Asia/Yangon | (UTC+06:30) Myanmar Time (Yangon) |
| Asia/Yekaterinburg | (UTC+05:00) Yekaterinburg Standard Time |
| Asia/Yerevan | (UTC+04:00) Armenia Standard Time (Yerevan) |
| Atlantic/Azores | (UTC-01:00) Azores Time |
| Atlantic/Bermuda | (UTC-04:00) Atlantic Time (Bermuda) |
| Atlantic/Canary | (UTC+00:00) Western European Time (Canary) |
| Atlantic/Cape_Verde | (UTC-01:00) Cape Verde Standard Time |
| Atlantic/Faroe | (UTC+00:00) Western European Time (Faroe) |
| Atlantic/Madeira | (UTC+00:00) Western European Time (Madeira) |
| Atlantic/Reykjavik | (UTC+00:00) Iceland Time |
| Atlantic/South_Georgia | (UTC-02:00) South Georgia Time |
| Atlantic/St_Helena | (UTC+00:00) St. Helena Time |
| Atlantic/Stanley | (UTC-03:00) Falkland Islands Standard Time (Stanley) |
| Australia/Adelaide | (UTC+09:30) Central Australia Time (Adelaide) |
| Australia/Brisbane | (UTC+10:00) Eastern Australia Time (Brisbane) |
| Australia/Broken_Hill | (UTC+09:30) Central Australia Time (Broken Hill) |
| Australia/Currie | (UTC+10:00) Eastern Australia Time (Currie) |
| Australia/Darwin | (UTC+09:30) Central Australia Time (Darwin) |
| Australia/Eucla | (UTC+08:45) Australian Central Western Standard Time (Eucla) |
| Australia/Hobart | (UTC+10:00) Eastern Australia Time (Hobart) |
| Australia/Lindeman | (UTC+10:00) Eastern Australia Time (Lindeman) |
| Australia/Lord_Howe | (UTC+10:30) Lord Howe Time |
| Australia/Melbourne | (UTC+10:00) Eastern Australia Time (Melbourne) |
| Australia/Perth | (UTC+08:00) Australian Western Standard Time (Perth) |
| Australia/Sydney | (UTC+10:00) Eastern Australia Time (Sydney) |
| Europe/Amsterdam | (UTC+01:00) Central European Time (Amsterdam) |
| Europe/Andorra | (UTC+01:00) Central European Time (Andorra) |
| Europe/Astrakhan | (UTC+04:00) Astrakhan Time |
| Europe/Athens | (UTC+02:00) Eastern European Time (Athens) |
| Europe/Belgrade | (UTC+01:00) Central European Time (Belgrade) |
| Europe/Berlin | (UTC+01:00) Central European Time (Berlin) |
| Europe/Bratislava | (UTC+01:00) Central European Time (Bratislava) |
| Europe/Brussels | (UTC+01:00) Central European Time (Brussels) |
| Europe/Bucharest | (UTC+02:00) Eastern European Time (Bucharest) |
| Europe/Budapest | (UTC+01:00) Central European Time (Budapest) |
| Europe/Busingen | (UTC+01:00) Central European Time (Busingen) |
| Europe/Chisinau | (UTC+02:00) Eastern European Time (Chisinau) |
| Europe/Copenhagen | (UTC+01:00) Central European Time (Copenhagen) |
| Europe/Dublin | (UTC+00:00) Ireland Time |
| Europe/Gibraltar | (UTC+01:00) Central European Time (Gibraltar) |
| Europe/Guernsey | (UTC+00:00) Guernsey Time |
| Europe/Helsinki | (UTC+02:00) Eastern European Time (Helsinki) |
| Europe/Isle_of_Man | (UTC+00:00) Isle of Man Time |
| Europe/Istanbul | (UTC+03:00) Turkey Time |
| Europe/Jersey | (UTC+00:00) Jersey Time |
| Europe/Kaliningrad | (UTC+02:00) Eastern European Time (Kaliningrad) |
| Europe/Kiev | (UTC+02:00) Eastern European Time (Kiev) |
| Europe/Kirov | (UTC+03:00) Kirov Time |
| Europe/Lisbon | (UTC+00:00) Western European Time (Lisbon) |
| Europe/Ljubljana | (UTC+01:00) Central European Time (Ljubljana) |
| Europe/London | (UTC+00:00) United Kingdom Time |
| Europe/Luxembourg | (UTC+01:00) Central European Time (Luxembourg) |
| Europe/Madrid | (UTC+01:00) Central European Time (Madrid) |
| Europe/Malta | (UTC+01:00) Central European Time (Malta) |
| Europe/Mariehamn | (UTC+02:00) Eastern European Time (Mariehamn) |
| Europe/Minsk | (UTC+03:00) Belarus Time |
| Europe/Monaco | (UTC+01:00) Central European Time (Monaco) |
| Europe/Moscow | (UTC+03:00) Moscow Time |
| Europe/Oslo | (UTC+01:00) Central European Time (Oslo) |
| Europe/Paris | (UTC+01:00) Central European Time (Paris) |
| Europe/Podgorica | (UTC+01:00) Central European Time (Podgorica) |
| Europe/Prague | (UTC+01:00) Central European Time (Prague) |
| Europe/Riga | (UTC+02:00) Eastern European Time (Riga) |
| Europe/Rome | (UTC+01:00) Central European Time (Rome) |
| Europe/Samara | (UTC+04:00) Samara Standard Time |
| Europe/San_Marino | (UTC+01:00) Central European Time (San Marino) |
| Europe/Sarajevo | (UTC+01:00) Central European Time (Sarajevo) |
| Europe/Saratov | (UTC+04:00) Saratov Time |
| Europe/Simferopol | (UTC+03:00) Simferopol Time |
| Europe/Skopje | (UTC+01:00) Central European Time (Skopje) |
| Europe/Sofia | (UTC+02:00) Eastern European Time (Sofia) |
| Europe/Stockholm | (UTC+01:00) Central European Time (Stockholm) |
| Europe/Tallinn | (UTC+02:00) Eastern European Time (Tallinn) |
| Europe/Tirane | (UTC+01:00) Central European Time (Tirane) |
| Europe/Ulyanovsk | (UTC+04:00) Ulyanovsk Time |
| Europe/Uzhgorod | (UTC+02:00) Eastern European Time (Uzhhorod) |
| Europe/Vaduz | (UTC+01:00) Central European Time (Vaduz) |
| Europe/Vatican | (UTC+01:00) Central European Time (Vatican) |
| Europe/Vienna | (UTC+01:00) Central European Time (Vienna) |
| Europe/Vilnius | (UTC+02:00) Eastern European Time (Vilnius) |
| Europe/Volgograd | (UTC+04:00) Volgograd Standard Time |
| Europe/Warsaw | (UTC+01:00) Central European Time (Warsaw) |
| Europe/Zagreb | (UTC+01:00) Central European Time (Zagreb) |
| Europe/Zaporozhye | (UTC+02:00) Eastern European Time (Zaporozhye) |
| Europe/Zurich | (UTC+01:00) Central European Time (Zurich) |
| Indian/Antananarivo | (UTC+03:00) East Africa Time (Antananarivo) |
| Indian/Chagos | (UTC+06:00) Indian Ocean Time (Chagos) |
| Indian/Christmas | (UTC+07:00) Christmas Island Time |
| Indian/Cocos | (UTC+06:30) Cocos Islands Time |
| Indian/Comoro | (UTC+03:00) East Africa Time (Comoro) |
| Indian/Kerguelen | (UTC+05:00) French Southern & Antarctic Time (Kerguelen) |
| Indian/Mahe | (UTC+04:00) Seychelles Time |
| Indian/Maldives | (UTC+05:00) Maldives Time |
| Indian/Mauritius | (UTC+04:00) Mauritius Standard Time |
| Indian/Mayotte | (UTC+03:00) East Africa Time (Mayotte) |
| Indian/Reunion | (UTC+04:00) Réunion Time |
| Pacific/Apia | (UTC+13:00) Samoa Time |
| Pacific/Auckland | (UTC+12:00) New Zealand Time |
| Pacific/Bougainville | (UTC+11:00) Bougainville Time |
| Pacific/Chatham | (UTC+12:45) Chatham Time |
| Pacific/Chuuk | (UTC+10:00) Chuuk Time |
| Pacific/Easter | (UTC-06:00) Easter Island Time |
| Pacific/Efate | (UTC+11:00) Vanuatu Standard Time (Efate) |
| Pacific/Enderbury | (UTC+13:00) Phoenix Islands Time (Enderbury) |
| Pacific/Fakaofo | (UTC+13:00) Tokelau Time |
| Pacific/Fiji | (UTC+12:00) Fiji Time |
| Pacific/Funafuti | (UTC+12:00) Tuvalu Time |
| Pacific/Galapagos | (UTC-06:00) Galapagos Time |
| Pacific/Gambier | (UTC-09:00) Gambier Time |
| Pacific/Guadalcanal | (UTC+11:00) Solomon Islands Time |
| Pacific/Guam | (UTC+10:00) Chamorro Standard Time (Guam) |
| Pacific/Honolulu | (UTC-10:00) Hawaii-Aleutian Time (Honolulu) |
| - | - |
| Pacific/Kiritimati | (UTC+14:00) Line Islands Time (Kiritimati)
| Pacific/Kosrae | (UTC+11:00) Kosrae Time |
| Pacific/Kwajalein | (UTC+12:00) Marshall Islands Time (Kwajalein) |
| Pacific/Majuro | (UTC+12:00) Marshall Islands Time |
| Pacific/Marquesas | (UTC-09:30) Marquesas Time |
| Pacific/Midway | (UTC-11:00) Samoa Standard Time (Midway) |
| Pacific/Nauru | (UTC+12:00) Nauru Time |
| Pacific/Niue | (UTC-11:00) Niue Time |
| Pacific/Norfolk | (UTC+11:00) Norfolk Island Time |
| Pacific/Noumea | (UTC+11:00) New Caledonia Standard Time (Noumea) |
| Pacific/Pago_Pago | (UTC-11:00) Samoa Standard Time (Pago Pago) |
| Pacific/Palau | (UTC+09:00) Palau Time |
| Pacific/Pitcairn | (UTC-08:00) Pitcairn Islands Time |
| Pacific/Pohnpei | (UTC+11:00) Ponape Time (Pohnpei) |
| Pacific/Port_Moresby | (UTC+10:00) Papua New Guinea Time (Port Moresby) |
| Pacific/Rarotonga | (UTC-10:00) Cook Islands Standard Time (Rarotonga) |
| Pacific/Saipan | (UTC+10:00) Chamorro Standard Time (Saipan) |
| Pacific/Tahiti | (UTC-10:00) Tahiti Time |
| Pacific/Tarawa | (UTC+12:00) Gilbert Islands Time (Tarawa) |
| Pacific/Tongatapu | (UTC+13:00) Tonga Standard Time (Tongatapu) |
| Pacific/Wake | (UTC+12:00) Wake Island Time |
| Pacific/Wallis | (UTC+12:00) Wallis & Futuna Time |
| UTC | (UTC) Coordinated Universal Time |
</details>

The difference between CoreCLR's TimeZones and Android's TimeZones is https://gist.github.com/mdh1418/acae237c7210551e249952b3ff4b2887
There are slight differences in 12 TimeZone DisplayNames, and there are 3 TimeZones found on Android not found on CoreCLR and 2 TimeZones found on CoreCLR and not found on Android, a great improvement over the previous difference before this change.
[AndroidFilteredIDsTwo_sort.txt](https://github.com/dotnet/runtime/files/7901923/AndroidFilteredIDsTwo_sort.txt)
[CoreCLRTimeZones_sort.txt](https://github.com/dotnet/runtime/files/7901925/CoreCLRTimeZones_sort.txt)

mdh1418 committed:
[libraries][Android] Filter out backward timezone IDs using tzlookup on Android

mdh1418 committed:
[tests] Add TimeZoneInfo Test to detect Backward TimeZones

danmoseley commented on src/libraries/System.Private.CoreLib/src/Resources/Strings.resx:
```diff
@@ -2593,6 +2593,9 @@
     <value>The Timer was already closed using an incompatible Dispose method.</value>
   </data>
+  <data name="InvalidOperation_TimeZoneLookupParsing" xml:space="preserve">
+    <value>Internal Error in parsing tzlookup file {1}.</value>
```
We try to put quotes around paths. Generally "

mdh1418 replied:
Done, thanks!

danmoseley commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -234,13 +235,94 @@ public AndroidTzData()
+                {
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_TimeZoneLookupParsing, lookupPath));
+                }
+                catch (OutOfMemoryException)
```
Just curious why out of memory?

mdh1418 replied:
I just looked through all the APIs being used in the method and looked at what possible exceptions could be thrown. We can omit this exception if its not necessary.

mdh1418 replied:
We are now catching all exceptions per https://github.com/dotnet/runtime/pull/64028#discussion_r789011873

am11 commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -234,13 +235,94 @@ public AndroidTzData()
+            // Once the timezone cache is populated with the IDs, we reference tzlookup id tags and
+            // remove IDs and their associated information (byteoffsets and lengths) from the current
+            // AndroidTzData instance.
+            private void filterBackwardIDs(string lookupPath)
```
```suggestion
            private void FilterBackwardIDs(string lookupPath)
```
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/docs/coding-guidelines/coding-style.md
> 13. We use PascalCasing for all method names, including local functions.

mdh1418 replied:
Thank you :)

am11 commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -234,13 +235,94 @@ public AndroidTzData()
+                            if (!(tzLookupLine = sr.ReadLine())!.TrimStart().StartsWith("<id"))
+                                continue;
+
+                            int idStart = tzLookupLine.IndexOf('>')+1;
+                            int idLength = tzLookupLine.IndexOf("</")-idStart;
```
```suggestion
                            int idStart = tzLookupLine.IndexOf('>') + 1;
                            int idLength = tzLookupLine.LastIndexOf("</") - idStart;
```

mdh1418 replied:
Thanks!

mdh1418 committed:
Address feedback

mdh1418 commented:
/azp run runtime-extra-platforms

MattGertz commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -234,13 +235,94 @@ public AndroidTzData()
+                        {
+                            if (!(tzLookupLine = sr.ReadLine())!.TrimStart().StartsWith("<id"))
+                                continue;
+
```
Is there not a standard XML parser that's used for XML files?  (Honest question; I have little familiarity with this codebase.)

steveisok replied:
No, not where the code is being changed.  The only option I know of other than this would be to interop with some native library. Would be overkill for this change in my opinion.

MattGertz replied:
Got it, thx.

MattGertz reviewed (approved):
LGTM

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -234,13 +235,94 @@ public AndroidTzData()
+                        string? tzLookupLine;
+                        while (sr.Peek() >= 0)
+                        {
+                            if (!(tzLookupLine = sr.ReadLine())!.TrimStart().StartsWith("<id"))
```
>StartsWith("<id") [](http://example.com/codeflow?start=77&length=17)

please use StringComparison.Ordinal here

mdh1418 replied:
Thanks!

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -234,13 +235,94 @@ public AndroidTzData()
+                        string? tzLookupLine;
+                        while (sr.Peek() >= 0)
+                        {
+                            if (!(tzLookupLine = sr.ReadLine())!.TrimStart().StartsWith("<id"))
```
>TrimStart [](http://example.com/codeflow?start=65&length=9)

I suggest to use the Span here to avoid the allocation

mdh1418 replied:
I'm not sure how, does using a span trim out the beginning white spaces?
Would it be better to just have 
`if (!(tzLookupLine = sr.ReadLine())!.Contains("<id", StringComparison.Ordinal))` instead ?

tarekgh replied:
`(tzLookupLine = sr.ReadLine())!.AsSpan()` should give you the span and then you can call Trim and StartWith on that span.

mdh1418 replied:
Done, thanks!

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -234,13 +235,94 @@ public AndroidTzData()
+
+                            int idStart = tzLookupLine.IndexOf('>') + 1;
+                            int idLength = tzLookupLine.LastIndexOf("</") - idStart;
+                            if (idStart == 0 || idLength < 0)
```
>==  [](http://example.com/codeflow?start=40&length=3)

make this condition <= instead of ==

mdh1418 replied:
Done. Is it possible for IndexOf to return a value less than -1?

tarekgh replied:
Yes, we document we return negative number. I recall there is some cases can return negative number other than -1.

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -234,13 +235,94 @@ public AndroidTzData()
+                                continue;
+
+                            int idStart = tzLookupLine.IndexOf('>') + 1;
+                            int idLength = tzLookupLine.LastIndexOf("</") - idStart;
```
>LastIndexOf [](http://example.com/codeflow?start=56&length=11)

use StringComparison.Ordinal

mdh1418 replied:
Thanks!

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -234,13 +235,94 @@ public AndroidTzData()
+                        _lengths = cleanLengths.ToArray();
+                    }
+                }
+                catch (FileNotFoundException)
```
>catch [](http://example.com/codeflow?start=16&length=5)

we should catch all exceptions and then ignore the filtering all together without any throwing. Getting the duplicated display names would be better than throwing and causing app failures

mdh1418 replied:
Ignore the filtering altogether, as in should we not even try to filter if `GetSystemTImeZones` is called a second time?
Should we log a warning here, and what is the standard way for logging that in CoreLib?

tarekgh replied:
> Ignore the filtering altogether, as in should we not even try to filter if GetSystemTImeZones is called a second time?

Yes. We try the filtering once and cache the result to use for any subsequent calls. If filtering fails, ignore it moving forward and don't try it again. It is unlikely to succeed if it fail first time. 

> Should we log a warning here, and what is the standard way for logging that in CoreLib?

I don't think we log that. Maybe we can have assert in the debug builds instead.

tarekgh replied:
One other note, we have the API TimeZoneInfo.ClearCachedData which clears all cached data. If this is called and we need to recreate the cache, we should do the filtering at that time.

mdh1418 replied:
The filtering is now based on an `_isBackwards` boolean array that is created and set when the IDs are populated. I believe the filtering still works after calling TimeZoneInfo.ClearCachedData

tarekgh replied:
> I believe the filtering still works after calling TimeZoneInfo.ClearCachedData

do you mean you recreate the `_isBackwards` array again or you are saying `_isBackwards` is created only once and used even after clearing the cache? 

Here is a scenario you can validate:

-  Call FindSystemTimeZoneUsingId("America/Los_Angeles")
- Call GetSystemTimeZones
- Validate the behavior that backward zones are filtered
- Clear the cache
- Call GetSystemTimeZones
- Validate the behavior one more time that backward zones are filtered

mdh1418 replied:
Its created whenever `_ids` are created (and `_byteOffsets` and `_lengths`), 
Using 
```
TimeZoneInfo LA = TimeZoneInfo.FindSystemTimeZoneById("America/Los_Angeles");
Console.WriteLine($"LA DisplayName: {LA.DisplayName}");
ReadOnlyCollection<TimeZoneInfo> tzCollection = TimeZoneInfo.GetSystemTimeZones();
HashSet<String> tzDisplayNames = new HashSet<String>();
foreach (TimeZoneInfo timezone in tzCollection)
{
    tzDisplayNames.Add(timezone.DisplayName);
}
Console.WriteLine($"{tzCollection.Count == tzDisplayNames.Count} tzCollection: {tzCollection.Count} tzDisplayNames: {tzDisplayNames.Count}");
TimeZoneInfo.ClearCachedData();
tzCollection = TimeZoneInfo.GetSystemTimeZones();
tzDisplayNames = new HashSet<String>();

foreach (TimeZoneInfo timezone in tzCollection)
{
    tzDisplayNames.Add(timezone.DisplayName);
}
Console.WriteLine($"{tzCollection.Count == tzDisplayNames.Count} tzCollection: {tzCollection.Count} tzDisplayNames: {tzDisplayNames.Count}");
```
Both evaluate to true

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -234,13 +235,94 @@ public AndroidTzData()
+                                cleanLengths.Add(_lengths[i]);
+                            }
+                        }
+                        _ids = cleanIDs.ToArray();
```
>_ids = cleanIDs.ToArray(); [](http://example.com/codeflow?start=24&length=26)

The question here, if someone try to manually create any of the backward zone (using FindSystemTimeZoneById) will fail. I think it will be better not fail it. 

Can we mark which Ids are backward and filter those only while enumerating the zones?

mdh1418 replied:
Good point, so instead of removing the backward timezones, we can instead create a parallel array of bools to track which are backward (where that array is updated by this function)?

mdh1418 replied:
Done, now there is an _isBackwards boolean array that is created and set when the id's and related byteoffset and lengths information are set as well.

mdh1418 commented:
The test to check for backward zones is failing on Windows/Linux, meaning that there are some duplicates on Windows/Linux. I cant easily reproduce on those platforms, but do we know if this is intended or if its also a bug? 
@tarekgh would you happen to know?
In the meantime, should they just be activeIssued on those platforms?

mdh1418 committed:
Address more feedback

Use StringComparison.Ordinal for strings
Catch and ignore exceptions
skip if indexof returns less than -1

tarekgh commented:
> The test to check for backward zones is failing on Windows/Linux, meaning that there are some duplicates on Windows/Linux. 

do you have the list of the zones which have the duplications?

mdh1418 committed:
[libraries][Android] Add isBackwards in lieu of removing timezone ids

mdh1418 committed:
Remove unneeded strings.resx additions

mdh1418 committed:
Temporarily show timezone displayNames in NoBackwardTimeZones

mdh1418 commented:
> do you have the list of the zones which have the duplications?

I didn't manage to build on Windows or Linux yet, I just added a temporary logging to the test to show the DisplayNames.

mdh1418 commented:
Looks like Windows passed this time, after changing from removing IDs to marking IDs as backwards. However the number of timezones in GetSystemTimeZones on Windows is 141, compared to 426 on Android.
Plus Linux's failing because there are GMT duplicates https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-64028-merge-6c95aab1c4be41fba8/System.Runtime.Tests/1/console.1303c0df.log?sv=2019-07-07&se=2022-02-10T02%3A11%3A58Z&sr=c&sp=rl&sig=sNOqWR0McQeR5eXzegrr%2B3id7d3ICCCPm6z7mg2Z2Tc%3D

mdh1418 committed:
Figure out backwards IDs on linux and maybe windows

mdh1418 committed:
[libraries] ActiveIssue NoBackwardTimeZones for Linux Arm64

mdh1418 commented:
/azp run runtime-extra-platforms

mdh1418 committed:
Cleanup

mdh1418 commented:
/azp run runtime-extra-platforms

mdh1418 commented:
@tarekgh I created an ActiveIssue for the NoBackwardTimeZones test on Linux arm64, can I get another look

mdh1418 committed:
Skip test with TestPlatforms instead

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs:
```diff
@@ -372,7 +428,15 @@ private void LoadEntryAt(Stream fs, long position, out string id, out int byteOf
             public string[] GetTimeZoneIds()
             {
-                return _ids;
+                List<string> nonBackwardsTZIDs = new List<string>();
```
nit: this code is ok but if you want to optimize more to avoid allocating the list, you can count the number of the filtered out  zones and then allocate the returned string[] with exact size. 

I'll leave this up to you if you want to optimize it or not.

mdh1418 replied:
Thanks for the suggestion! I looped through the `_isBackwards` array one extra time to count the number to allocate. It seems a bit weird to loop twice, but having a counter for the `nonBackwardsTZIDs` in the class would be allocating for an `int` for the entire lifetime of the class, and thats worse than just the lifetime of `GetTimeZoneIds` right?

Out of curiosity, what specifically are we avoiding in allocating the list? I am seeing that a list allocates space for a 2<sup>n</sup> array where its doubled each time it runs out of space. Are we avoiding that worst case 2<sup>n-1</sup>-1 space allocation or are we mainly avoiding the extra variables being allocated through the `List<T>` object?

grendello replied:
On mobile it's a very good idea to avoid any unnecessary allocations (on desktop/server too, but less critical) as it not only conserves resources but it also puts a lower pressure on the GC by generating less objects to scan.

tarekgh commented on src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:
```diff
@@ -2921,6 +2921,20 @@ public static void AdjustmentRuleBaseUtcOffsetDeltaTest()
         }
 
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/64111", TestPlatforms.Linux)]
```
>Linux [](http://example.com/codeflow?start=85&length=5)

I assume this test case will still run on Android. right?

tarekgh reviewed (approved):
Thanks @mdh1418  for getting this issue fixed.

mdh1418 committed:
Optimize further by not allocating a list for GetTimeZoneIDs

mdh1418 commented:
/azp run runtime-extra-platforms

mdh1418 committed:
Fix indexing

mdh1418 commented:
/azp run runtime-extra-platforms

mdh1418 commented:
/backport release/6.0-maui

mdh1418 commented:
/backport to release/6.0-maui

steveisok commented:
/backport to release/6.0-maui
